src/battle_ai_switch_items.c

```diff
@@ -228,44 +217,177 @@ static bool8 FindMonThatAbsorbsOpponentsMove(void)
     return FALSE;
 }
 
-static bool8 ShouldSwitchIfNaturalCure(void)
+static bool8 ShouldSwitchIfGameStatePrompt(void)
 {
-    if (!(gBattleMons[gActiveBattler].status1 & STATUS1_SLEEP))
-        return FALSE;
-    if (AI_GetAbility(gActiveBattler) != ABILITY_NATURAL_CURE)
-        return FALSE;
-    if (gBattleMons[gActiveBattler].hp < gBattleMons[gActiveBattler].maxHP / 2)
-        return FALSE;
+    bool8 switchMon = FALSE;
+    u16 monAbility = AI_GetAbility(gActiveBattler);
+    u16 holdEffect = AI_GetHoldEffect(gActiveBattler);
+    u8 opposingPosition = BATTLE_OPPOSITE(GetBattlerPosition(gActiveBattler));
+    u8 opposingBattler = GetBattlerAtPosition(opposingPosition);
+    s32 moduloChance = 4; //25% Chance Default
+    s32 chanceReducer = 1; //No Reduce default. Increase to reduce
+
+
+    if (AnyStatIsRaised(gActiveBattler))
+        chanceReducer = 5; // Reduce switchout probability by factor of 5 if setup
 
-    if ((gLastLandedMoves[gActiveBattler] == MOVE_NONE
-      || gLastLandedMoves[gActiveBattler] == MOVE_UNAVAILABLE)
-     && Random() & 1)
+    //Perish Song
+    if (gStatuses3[gActiveBattler] & STATUS3_PERISH_SONG
+        && gDisableStructs[gActiveBattler].perishSongTimer == 0
+        && monAbility != ABILITY_SOUNDPROOF)
+        switchMon = TRUE;
+
+    if (AI_THINKING_STRUCT->aiFlags & AI_FLAG_SMART_SWITCHING)
     {
-        *(gBattleStruct->AI_monToSwitchIntoId + gActiveBattler) = PARTY_SIZE;
-        BtlController_EmitTwoReturnValues(BUFFER_B, B_ACTION_SWITCH, 0);
-        return TRUE;
+        //Yawn
+        if (gStatuses3[gActiveBattler] & STATUS3_YAWN
+            && AI_CanSleep(gActiveBattler, monAbility)
+            && gBattleMons[gActiveBattler].hp > gBattleMons[gActiveBattler].maxHP / 3)
+        {
+            switchMon = TRUE;
+
+            //ToDo:
+                //Double Battles
+                //Add logic checking to see if effected by yawn & ally wants to switch out to a pokemon that will set Misty or Electric Terrain
+
+            //Check if Active Pokemon can KO opponent instead of switching
+            //Will still fall asleep, but take out opposing Pokemon first
+            if (AiExpectsToFaintPlayer())
+                switchMon = FALSE;
+
+            //Checks to see if active Pokemon can do something against sleep
+            if (monAbility == (ABILITY_NATURAL_CURE | ABILITY_SHED_SKIN | ABILITY_EARLY_BIRD)
```

Comment:
you can also check if your partner is going to use misty/electric terrain this turn, but checking if they have the move below may also be sufficient.

`IsDoubleBattle() && IsBattlerAlive(partner) && AI_DATA->partnerMove && gBattleMoves[AI_DATA->partnerMove].effect == EFFECT_MISTY_TERRAIN`, for example